icu4c/source/common/unistr.cpp

```diff
@@ -1947,6 +1947,11 @@ UnicodeString::cloneArrayIfNeeded(int32_t newCapacity,
       growCapacity = US_STACKBUF_SIZE;
     }
 
+    // The newCapacity is too large.
+    if (newCapacity > growCapacity) {
+      return false;
+    }
```

Comment:
let's first check for valid arguments,
then check the allocation size for the named max constant
```suggestion
    } else if(newCapacity > growCapacity) {
      return false;  // bad inputs
    }

    if(growCapacity > kMaxCapacity) {
      return false;
    }
```